static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,340 @@
 <?xml version="1.0"?>
 <macker>
-    <ruleset name="Test cases should not be dependent on each other">
-        <var name="testCasesPackage" value="teammates.test.cases" />
-        <access-rule>
-            <deny>
-                <from class="${testCasesPackage}.**" />
-                <to class="${testCasesPackage}.**" />
+    <ruleset name="Architectural rules for production code">
+        <pattern name="productionCode">
+            <exclude class="teammates.test.**" />
+            <exclude class="teammates.client.**" />
+        </pattern>
+        <ruleset name="PackageDiagram: UI should not touch Storage">
+            <access-rule>
+                <deny>
+                    <from class="teammates.ui.**" />
+                    <to class="teammates.storage.**" />
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="PackageDiagram: Logic should not touch UI">
+            <access-rule>
+                <deny>
+                    <from class="teammates.logic.**" />
+                    <to class="teammates.ui.**" />
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="PackageDiagram: Storage should not touch Logic">
+            <access-rule>
+                <deny>
+                    <from class="teammates.storage.**" />
+                    <to class="teammates.logic.**" />
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="PackageDiagram: Storage should not touch UI">
+            <access-rule>
+                <deny>
+                    <from class="teammates.storage.**" />
+                    <to class="teammates.ui.**" />
+                </deny>
+            </access-rule>
+        </ruleset>
+        
+        <ruleset name="UiComponent: Only *Action can touch Logic API">
+            <access-rule>
+                <deny>
+                    <from class="teammates.ui.**" />
+                    <to class="teammates.logic.**" />   
+                    <allow>
+                        <from>
+                            <include class="teammates.ui.controller.**" />
+                            <include class="teammates.ui.automated.**" />
+                        </from>
+                        <to class="teammates.logic.api.**" />
+                    </allow>
+                    <allow>
+                        <!-- to import sample data, need to be changed -->
+                        <from class="teammates.ui.controller.AdminInstructorAccountAddAction" />
+                        <to class="teammates.logic.backdoor.BackDoorLogic" />
+                    </allow>    
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="UiComponent: Templates are only used in Pagedata">
+            <access-rule>
+                <deny>
+                    <from pattern="productionCode" />
```

Comment:
What if `pagedata` is used outside `ui` package? Currently, only `storage` and `logic` cannot touch UI.